crankshaft/compiler.md

```diff
@@ -0,0 +1,281 @@
+  - if assumptions are violated again, deoptimized again and start over 
+- too many deoptimizations cause function to be sent to *deoptimization hell*
+  - considered not optimizable and no optimization is **ever** attempted again
+- certain constructs like `try/catch` are considered not optimizable and functions containing it go straight to
+  *deoptimization hell* due to **bailout** [watch](http://youtu.be/UJPdhx5zTaw?t=35m23s)
+
+None of this can be diagnosed with Chrome Devtools at this point.
+
+### Causes for Deoptimization
+
+#### Modifying Object Shape
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=21m00s)
+
+- added fields (order matters) to object generate id of hidden class
+- adding more fields later on generates new class id which results in code using Point that now gets Point' to be
+  deoptimized
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=21m45s)
+[watch](http://youtu.be/UJPdhx5zTaw?t=12m18s)
+
+```js
+function Point(x, y) {
+  this.x = x;
+  this.y = y;
+}
+
+var p = new Point(1, 2); // => hidden Point class created
+
+// ....
+
+p.z = 3;                 // => another hidden class (Point') created
+```
+
+- `Point` class created, code still deoptimized
+- functions that have `Point` argument are optimized
+- `z` property added which causes `Point'` class to be created
+- functions that get passed `Point'` but were optimized for `Point` get deoptimized
+- later functions get optimized again, this time supporting `Point` and `Point'` as argument
+- [detailed explanation](http://v8-io12.appspot.com/index.html#30)
+
+##### Considerations
+
+- avoid hidden class changes
+- initialize all members in **constructor function** and **in the same order**
+
+## Efficiently Representing Values and Tagging
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=15m35s) | [slide](http://v8-io12.appspot.com/index.html#34)
+
+- v8 passes around 32bit numbers to represent all values
+- bottom bit reserved as tag to signify if value is a SMI (small integer) or a pointer to an object
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=10m05ss) | [slide](http://v8-io12.appspot.com/index.html#35)
+
+```
+| object pointer              | 1 |
+
+or
+
+| 31-bit-signed integer (SMI) | 0 |
+```
+
+- numbers bigger than 31 bits are boxed 
+- stored inside an object referenced via a pointer 
+- adds extra overhead (at a minimum an extra lookup)
+
+### Considerations
+
+- prefer SMIs for numeric values whenever possible
+
+## Arrays
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=17m25s) | [slide](http://v8-io12.appspot.com/index.html#38)
+
+v8 has two methods for storing arrays.
+
+### Fast Elements
+
+- compact keysets
+- linear storage buffer
+
+#### Characteristics
+
+- contiguous (non-sparse) 
+- `0` based
+- smaller than 64K
+
+### Dictionary Elements
+
+- hash table storage
+- slow access
+
+#### Characteristics
+
+- sparse
+- large
+
+### Double Array Unboxing
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=20m20s) | [slide](http://v8-io12.appspot.com/index.html#45)
```

Comment:
Alternate, more recent presentation: https://www.youtube.com/watch?v=m9cTaYI95Zc Slides are linked in the description

Reply:
Video and post were already linked in resources of the updated data-types.md.
I added the slides as well now.

However I couldn't see a part in the video that focuses particularly on double array unboxing the way that the linked (out of date) video + slide does. Did I miss something and if so please point me to the slide or time in the up to date video.